### ISSUE TITLE
Proyecto Final

### ISSUE THREAD

**Issue by DrGuaracha** [DrGuaracha/Dataxperience#1]
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import r2_score
import matplotlib.pyplot as plt
import seaborn as sns

# 1. Carga y limpieza de datos
try:
    df = pd.read_csv('vehiculos.csv')
    print("Datos cargados desde 'vehiculos.csv'.")

except FileNotFoundError:
    print("Error: 'vehiculos.csv' no encontrado. Creando un DataFrame de ejemplo.")
    # Crea un DataFrame de ejemplo para que el código funcione
    df = pd.DataFrame({
        'precio': [15000, 25000, 35000, 12000, 8000, 20000, 45000, 18000, 5000, 7500],
        'año': [2018, 2020, 2022, 2015, 2010, 2021, 2023, 2019, 2008, 2012],
        'kilometraje': [50000, 25000, 15000, 80000, 120000, np.nan, 35000, 60000, 150000, 95000],
        'marca': ['Ford', 'Tesla', 'Mercedes-Benz', 'Toyota', 'Nissan', 'Ford', 'Tesla', 'Mercedes Benz', 'Nissan', 'Ford'],
        'modelo': ['Fiesta', 'Model 3', 'Clase C', 'Corolla', 'Versa', 'Fiesta', 'Model 3', 'Clase C', 'Versa', 'Focus'],
        'tipo_de_motor': ['gasolina', 'eléctrico', 'gasolina', 'gasolina', 'gasolina', 'gasolina', 'eléctrico', 'gasolina', np.nan, 'gasolina'],
        'color': ['rojo', 'blanco', 'negro', 'plata', 'azul', 'rojo', 'blanco', 'negro', 'azul', 'plata']
    })

# Manejo de nulos y limpieza básica
df['kilometraje'].fillna(df['kilometraje'].median(), inplace=True)
df.dropna(subset=['tipo_de_motor'], inplace=True)
df.drop_duplicates(inplace=True)

# Filtrar valores extremos
df = df[(df['precio'] > 1000) & (df['año'] > 1990)]

# Normalizar texto
for col in ['marca', 'modelo', 'tipo_de_motor', 'color']:
    df[col] = df[col].astype(str).str.lower().str.strip().replace({'mercedes-benz': 'mercedes benz'})

print(f"Limpieza completa. {len(df)} filas restantes.")

# 2. Preparación de datos para el modelo
X = df.drop('precio', axis=1)
y = df['precio']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

variables_numericas = ['año', 'kilometraje']
variables_categoricas = ['marca', 'modelo', 'tipo_de_motor', 'color']

preprocesador = ColumnTransformer(
    transformers=[
        ('num', StandardScaler(), variables_numericas),
        ('cat', OneHotEncoder(handle_unknown='ignore'), variables_categoricas)
    ],
    remainder='passthrough'
)

X_train_processed = preprocesador.fit_transform(X_train)
X_test_processed = preprocesador.transform(X_test)

# 3. Entrenamiento y evaluación del modelo
modelo = RandomForestRegressor(n_estimators=100, random_state=42, n_jobs=-1)
modelo.fit(X_train_processed, y_train)

y_pred = modelo.predict(X_test_processed)
r2 = r2_score(y_test, y_pred)

print(f"\nModelo entrenado. Coeficiente R-cuadrado: {r2:.2f}")

# 4. Visualización de resultados
plt.figure(figsize=(10, 6))
sns.scatterplot(x=y_test, y=y_pred, alpha=0.6)
plt.plot([y.min(), y.max()], [y.min(), y.max()], 'r--', lw=2, label='Predicción Perfecta')
plt.title('Predicciones del Modelo vs. Precios Reales')
plt.xlabel('Precios Reales')
plt.ylabel('Predicciones del Modelo')
plt.legend()
plt.grid(True)
plt.show()

# 5. Ejemplo de uso del modelo
vehiculo_nuevo = pd.DataFrame([{
    'marca': 'ford',
    'modelo': 'fiesta',
    'año': 2021,
    'kilometraje': 30000,
    'tipo_de_motor': 'gasolina',
    'color': 'rojo'
}])

vehiculo_procesado = preprocesador.transform(vehiculo_nuevo)
precio_predicho = modelo.predict(vehiculo_procesado)

print(f"\nEl precio predicho para el vehículo es: ${precio_predicho[0]:,.2f}")
